src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt

```diff
@@ -0,0 +1,184 @@
+import io.ktor.generator.cli.utils.*
+import io.ktor.generator.configuration.json.*
+import io.ktor.http.*
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.toKString
+import kotlinx.cinterop.usePinned
+import kotlinx.coroutines.runBlocking
+import platform.posix.*
+import kotlin.system.exitProcess
+
+internal fun getEnv(name: String): String? = getenv(name)?.toKString()
+
+abstract class KtorInstaller(private val client: HttpClient, private val ktorBackendHost: String) {
+    abstract val rootKtorDirName: String
+    abstract val jdkDownloadUrl: String
+
+    abstract val jdkArchiveName: String
+
+    abstract fun unpackJdk(archive: File, outputDir: Directory)
+
+    abstract fun isGradleWrapper(file: File): Boolean
+
+    protected val ktorRootDir: Directory by lazy { Directory.home().createDirIfNeeded(rootKtorDirName) }
+    protected val ktorRcFile: File by lazy { ktorRootDir.createFileIfNeeded(KTOR_RC_FILENAME) }
+
+    private fun runGradle(gradleFile: File, task: String, javaHome: String, args: List<String> = emptyList()) {
+        setenv(JAVA_HOME, javaHome, 1)
+        addExecutablePermissions(gradleFile)
+        runProcess("${gradleFile.path} $task ${args.joinToString(" ")}")
+    }
+
+    private fun getRcProperties(): Map<String, String> = ktorRcFile.readLines().associate {
+        val (name, value) = it.split(" ")
+        name to value
+    }
+
+    private fun addRcProperty(name: String, value: String) {
+        ktorRcFile.writeText("$name $value\n")
+    }
+
+    private fun getRcProperty(name: String): String? = getRcProperties()[name]
+
+    private fun createAdHockJdkDir(): Directory = ktorRootDir.createDirIfNeeded(JDK_INSTALLED_DIR_PATH)
+
+    private fun initKtorRootIfAbsent() {
+        assert(ktorRootDir.exists())
+        assert(ktorRcFile.exists())
+    }
+
+    private fun findCustomJdk(): Directory? =
+        ktorRootDir
+            .content()
+            .filterIsInstance<Directory>()
+            .find { it.name == JDK_INSTALLED_DIR_PATH }
+            ?.subdir(JAVA_CONTENTS)
+            ?.subdir(JAVA_CONTENTS_HOME)
+
+    private fun hasAdHockJdk(): Boolean = findCustomJdk() != null
+
+    private fun getJavaHome(): String? = getEnv(JAVA_HOME)
+
+    private fun hasJavaHome11(): Boolean {
+        val javaHome = getJavaHome() ?: return false
+        return javaHome.contains("11")
+    }
+
+    private fun hasJdk(): Boolean = getRcProperty(JAVA_HOME) != null || hasAdHockJdk() || hasJavaHome11()
+
+    private fun installJdkIfAbsent() {
+        initKtorRootIfAbsent()
+        if (hasJdk()) {
+            if (getRcProperty(JAVA_HOME) == null) {
+                addRcProperty(JAVA_HOME, findCustomJdk()?.path ?: getJavaHome()!!)
+            }
+            return
+        }
+
+        val jdkArchiveFile = Directory.home().createFileIfNeeded(jdkArchiveName)
+
+        println("JDK 11 not found.\nJdk download path:$jdkDownloadUrl\nDownloading JDK-11 from server, please wait...")
```

Comment:
We should extract such string constants in a separate file and consider localization in the future.

It's hard to proofread it right now